### PR TITLE
Update Function CRD to support lambda code versioning

### PR DESCRIPTION
**Pull request by m00g3n**
**Description**
Update Function CRD to support lambda code versioning.

Changes proposed in this pull request:
- Function CRD changes to support lambda code versioning.

**Related issue(s)**

See also: https://github.com/kyma-project/kyma/issues/8959, https://github.com/kyma-project/kyma/issues/5752#issuecomment-648069580

### CODE REVIEW

File: components/function-controller/pkg/apis/serverless/v1alpha1/function_types.go [hunk 1]
```diff
@@ -27,6 +36,10 @@ type FunctionSpec struct {
 
 	// +optional
 	Labels map[string]string `json:"labels,omitempty"`
+
+	SourceType SourceType `json:"type"`
```
Comment (tgorgol) [inline components/function-controller/pkg/apis/serverless/v1alpha1/function_types.go:40]: Shouldn't this field be optional for the backward compatibility?

File: components/function-controller/pkg/apis/serverless/v1alpha1/function_types.go [hunk 2]
```diff
@@ -5,6 +5,15 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+type SourceType string
+
+type Runtime string
```
Comment (tgorgol) [inline components/function-controller/pkg/apis/serverless/v1alpha1/function_types.go:10]: `// +kubebuilder:validation:Enum=...` marker could be added to ensure validation on the API level.

File: components/function-controller/pkg/apis/serverless/v1alpha1/function_types.go [hunk 3]
```diff
@@ -79,6 +94,15 @@ type Condition struct {
 // FunctionStatus defines the observed state of FuncSONPath: .status.phase
 type FunctionStatus struct {
 	Conditions []Condition `json:"conditions,omitempty"`
+	Repository `json:",inline,omitempty"`
```
Comment (aerfio) [inline components/function-controller/pkg/apis/serverless/v1alpha1/function_types.go:99]: do we want this info in every function? shouldn't it be a pointer if it's not needed everywhere?

File: components/function-controller/pkg/apis/serverless/v1alpha1/function_types.go (hunk 2 above)
Comment (aerfio) [reply components/function-controller/pkg/apis/serverless/v1alpha1/function_types.go]: yeah, same for sourcetype

File: components/function-controller/pkg/apis/serverless/v1alpha1/function_types.go (hunk 1 above)
Comment (aerfio) [reply components/function-controller/pkg/apis/serverless/v1alpha1/function_types.go:42]: yeah, please make it a pointer and add kubebuilder markers

File: components/function-controller/pkg/apis/serverless/v1alpha1/function_types.go [hunk 4]
```diff
@@ -89,7 +113,7 @@ type FunctionStatus struct {
 // +kubebuilder:printcolumn:name="Running",type="string",JSONPath=".status.conditions[?(@.type=='Running')].status"
 // +kubebuilder:printcolumn:name="Version",type="integer",JSONPath=".metadata.generation"
 // +kubebuilder:printcolumn:name="Age",type="date",JSONPath=".metadata.creationTimestamp"
-
+// +kubebuilder:storageversion
```
Comment (aerfio) [inline components/function-controller/pkg/apis/serverless/v1alpha1/function_types.go:118]: it's not needed https://book.kubebuilder.io/multiversion-tutorial/api-changes.html#storage-versions

+ there needs to be 1 line of space between the markers and the struct as far as I remember

File: components/function-controller/pkg/apis/serverless/v1alpha1/function_types.go (hunk 3 above)
Comment (m00g3n) [reply components/function-controller/pkg/apis/serverless/v1alpha1/function_types.go:99]: Could you please help me understand the what do you mean? 
- in a non git type function we will get e.g. 
```
...
Status:
  Conditions:
    Last Transition Time:  2020-07-28T11:06:09Z
    Message:               Deployment demo-57vkq is not ready yet
    Reason:                DeploymentWaiting
    Status:                Unknown
    Type:                  Running
    Last Transition Time:  2020-07-28T11:06:07Z
    Message:               Job demo-build-8t8zq finished
    Reason:                JobFinished
    Status:                True
    Type:                  BuildReady
    Last Transition Time:  2020-07-28T11:05:54Z
    Message:               ConfigMap demo-thlxf created
    Reason:                ConfigMapCreated
    Status:                True
    Type:                  ConfigurationReady
  Source:                  module.exports = {
    main: function(event, context) {
      return 'Hello World'
    }
  }
```
- in git type function we will get e.g.
```
...
Status:
  Base Dir:  /
  Branch:    master
  Commit:    2e97a97079e2a41ab794d77fe1ccdc46d76e71db
  Conditions:
    Last Transition Time:  2020-07-28T11:13:49Z
    Message:               Job newone-build-tzjwz is still in progress
    Reason:                JobRunning
    Status:                Unknown
    Type:                  BuildReady
    Last Transition Time:  2020-07-28T11:13:46Z
    Message:               Sources newone updated
    Reason:                SourceUpdated
    Status:                True
    Type:                  ConfigurationReady
  Dockerfile:              deploy/Dockerfile
  Source:                  https://github.com/pPrecel/public-gitops
```

Comment (aerfio) [reply components/function-controller/pkg/apis/serverless/v1alpha1/function_types.go:99]: Ok, I misunderstood how it works, I thought that you just need a pointer + "omitempty" combo to achive that

**Pull request merged**